proxy/pkg/zdmproxy/streamidmapper.go

```diff
@@ -0,0 +1,62 @@
+package zdmproxy
+
+import (
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"sync"
+)
+
+// StreamIdMapper is used to map the incoming stream ids from the client/driver to internal ids managed by the proxy
+// This is required because we also generate requests inside the proxy that go to the cluster through the same connection,
+// hence they must have non-conflicting ids with user's requests.
+type StreamIdMapper interface {
+	GetNewIdFor(streamId int16) (int16, error)
+	ReleaseId(syntheticId int16) (int16, error)
+}
+
+type streamIdMapper struct {
+	sync.Mutex
+	idMapper   map[int16]int16
+	clusterIds chan int16
+}
+
+func NewStreamIdMapper(maxStreamIds int) StreamIdMapper {
+	idMapper := make(map[int16]int16)
+	streamIdsQueue := make(chan int16, maxStreamIds)
+	for i := int16(0); i < int16(maxStreamIds); i++ {
+		streamIdsQueue <- i
+	}
+	return &streamIdMapper{
+		idMapper:   idMapper,
+		clusterIds: streamIdsQueue,
+	}
+}
+
+func (sim *streamIdMapper) GetNewIdFor(streamId int16) (int16, error) {
+	sim.Lock()
+	defer sim.Unlock()
+	select {
+	case id, ok := <-sim.clusterIds:
+		if ok {
+			sim.idMapper[id] = streamId
```

Comment:
We should check if there already is an entry for this id because if there is then it's a stream id collision (probably a bug) and we need to return an error instead. 

I think I originally mentioned we could maybe remove the check but I think it does make sense to check it. If we had the check I think you would have noticed a bug in the control connection.

Reply:
Alright, I've added the check back.